File: go-ds-and-algorithms/list/linked_list.go
package list

type LinkedList struct {
	Head   *Node
	Length int
}

func NewLinkedList() LinkedList {
	return LinkedList{
		Head:   &Node{},
		Length: 0,
	}
}

func (list *LinkedList) Read(index int) *Node {
	curNode := list.Head.Next
	curIndex := 0
	for curNode != nil && curIndex < index {
		curNode = curNode.Next
		curIndex++
	}
	if curNode == nil {
		return nil
	}
	return curNode
}

func (list *LinkedList) Lookup(val int) *Node {
	curNode := list.Head.Next
	for curNode != nil {
		if val == curNode.Val {
			return curNode
		}
		curNode = curNode.Next
	}

	return nil
}

func (list *LinkedList) Insert(index int, val int) bool {
	if index < 0 || index > list.Length {
		return false
	}

	curIndex := 0
	curNode := list.Head
	for curIndex < index {
		curNode = curNode.Next
		curIndex++
	}
	node := Node{Val: val}
	node.Next = curNode.Next
	curNode.Next = &node
	list.Length++
	return true
}

func (list *LinkedList) Remove(index int) bool {
	if index < 0 || index > list.Length-1 {
		return false
	}

	curIndex := 0
	curNode := list.Head
	for curIndex < index {
		curNode = curNode.Next
		curIndex++
	}
	list.Length--
	return true
}

File: java-ds-and-algorithms/src/main/java/list/dual/LinkedList.java
package list.dual;

import lombok.Data;

import java.util.Objects;

/**
 * Usage: 双向链表 <br/>
 * Date: 2023/5/22 16:03 <br/>
 *
 * @author <a href="mailto:hanzhang2566@foxmail.com">hanzhang</a>
 */
@Data
public class LinkedList {
    /**
     * 长度
     */
    private int length;

    private Node head;

    public LinkedList(Node head) {
        this.head = head;
        length = 0;
    }

    public Node read(int index) {
        if (index < 0 || index > length) {
            return null;
        }
        int curIndex = 0;
        Node curNode = head.getNext();
        while (curIndex < index) {
            curNode = curNode.getNext();
            curIndex++;
        }
        return curNode;
    }

    public Node lookup(int val) {
        Node curNode = head.getNext();
        while (Objects.nonNull(curNode)) {
            if (curNode.getVal() == val) {
                return curNode;
            }
            curNode = curNode.getNext();
        }
        return null;
    }

    public boolean insert(int index, int val) {
        if (index < 0 || index > length) {
            return false;
        }
        int curIndex = 0;
        Node curNode = head;
        while (curIndex < index) {
            curNode = curNode.getNext();
            curIndex++;
        }
        Node next = curNode.getNext();
        Node node = new Node(val);
        if (Objects.isNull(next)) {
            curNode.setNext(node);
            node.setPrevious(curNode);
        } else {
            node.setNext(curNode.getNext());
            node.setPrevious(curNode);

            curNode.getNext().setPrevious(node);
            curNode.setNext(node);
        }
        length++;
        return true;
    }

    public boolean remove(int index) {
//        if (index < 0 || index < length) {
//            return false;
//        }
//        Node curNode = head;
//        int curIndex = 0;
//        while (curIndex < index) {
//            curNode = curNode.getNext();
//            curIndex++;
//        }
//        Node next = curNode.getNext();
//        if (Objects.isNull(next)) {
//            curNode.setNext(null);
//        } else {
//            curNode.getNext().getNext().setPrevious(curNode);
//            curNode.setNext(curNode.getNext().getNext());
//        }
        // FIXME
        return true;
    }

    @Override
    public String toString() {
        Node curNode = head.getNext();
        int count = 1;
        StringBuilder sb = new StringBuilder("{");
        while (Objects.nonNull(curNode)) {
            int val = curNode.getVal();
            sb.append(count);
            sb.append(":");
            sb.append(val);
            sb.append(",");
            count++;
            curNode = curNode.getNext();
        }
        sb = new StringBuilder(sb.substring(0, sb.length() - 1)).append("}");
        return sb.toString();
    }
}
